### PR TITLE
`get_required_signatures` API returns owner keys

### PR DESCRIPTION
**Pull request by abitmore**
PR for #567, #197, #200.

Related: #496, #497.

### CODE REVIEW

**Comment by oxarbitrage** [comment]
getting error when running the new test case:

```
root@alfredo:~/pull570/bitshares-core/tests# ./chain_test -t database_api_tests/get_required_signatures_owner_or_active -l all
Random number generator seeded to 1515699270
GRAPHENE_TESTING_GENESIS_TIMESTAMP is 1431700000
Running 1 test case...
Entering test suite "Master Test Suite"
Entering test suite "database_api_tests"
Entering test case "get_required_signatures_owner_or_active"
2070452ms th_a       db_management.cpp:201         close                ] Database close unexpected exception: {"code":10,"name":"assert_exception","message":"Assert Exception","stack":[{"context":{"level":"error","file":"index.hpp","line":111,"method":"get","hostname":"","thread_name":"th_a","timestamp":"2018-01-11T19:34:30"},"format":"maybe_found != nullptr: Unable to find Object","data":{"id":"2.1.0"}}]}
unknown location(0): fatal error in "get_required_signatures_owner_or_active": std::runtime_error: locale::facet::_S_create_c_locale name not valid
Leaving test case "get_required_signatures_owner_or_active"; testing time: 19430mks
Leaving test suite "database_api_tests"
Leaving test suite "Master Test Suite"

*** 1 failure detected in test suite "Master Test Suite"
root@alfredo:~/pull570/bitshares-core/tests# 
```

**Comment by abitmore** [comment]
@oxarbitrage something might be wrong with your environment. I've seen that error message before, but forgot how to get around it.

File: libraries/chain/protocol/transaction.cpp [hunk 1]
```diff
@@ -334,7 +334,7 @@ set<public_key_type> signed_transaction::get_required_signatures(
    for( auto& owner : required_owner )
       s.check_authority( get_owner( owner ) );
    for( auto& active : required_active )
-      s.check_authority( active  );
+      s.check_authority( active ) || s.check_authority( get_owner( active ) );
```
Comment (abitmore) [inline libraries/chain/protocol/transaction.cpp]: Perhaps better check `owner` first, so when the client can sign with owner key, no need to sign with active key at same time? (See #197)

File: libraries/app/database_api.cpp [hunk 2]
```diff
@@ -1840,9 +1840,6 @@ set<public_key_type> database_api_impl::get_potential_signatures( const signed_t
          const auto& auth = id(_db).active;
          for( const auto& k : auth.get_keys() )
             result.insert(k);
-         // Also insert owner keys since owner can authorize a trx that requires active only
-         for( const auto& k : id(_db).owner.get_keys() )
-            result.insert(k);
```
Comment (pmconrad) [inline libraries/app/database_api.cpp]: Why is this removed? According to the documentation,
> This method will return the set of all public keys that could possibly sign for a given transaction.

That includes the owner keys of "active" authorities, doesn't it?

**Comment by pmconrad** [comment]
Also, if the general rule is that owner can always sign in place of active, then the recursion in sign_state::check_authority must also check for the owner auth (https://github.com/abitmore/bitshares-core/blob/97fe7208e62ef710ed16badd2404d738851965fd/libraries/chain/protocol/transaction.cpp#L199).

File: libraries/app/database_api.cpp (hunk 2 above)
Comment (abitmore) [reply libraries/app/database_api.cpp]: Owner keys **of immediately required accounts** are implicitly included due to another change in same commit:
```
-      s.check_authority( active  );
+      s.check_authority( active ) || s.check_authority( get_owner( active ) );
```

Owner keys **of non-immediately required accounts** are unable to authorize the transaction due to [the behavior you mentioned below](https://github.com/bitshares/bitshares-core/pull/570#issuecomment-357867908).

**Comment by abitmore** [comment]
>Also, if the general rule is that owner can always sign in place of active, then the recursion in sign_state::check_authority must also check for the owner auth (https://github.com/abitmore/bitshares-core/blob/97fe7208e62ef710ed16badd2404d738851965fd/libraries/chain/protocol/transaction.cpp#L199).

@pmconrad yes, you're correct. Perhaps also here https://github.com/abitmore/bitshares-core/blob/97fe7208e62ef710ed16badd2404d738851965fd/libraries/chain/protocol/transaction.cpp#L164 . Since it requires a hard fork to fix it, I won't touch it this time. I think it's best to discuss it in a new ticket.

File: libraries/app/database_api.cpp (hunk 2 above)
Comment (pmconrad) [reply libraries/app/database_api.cpp]: You're right, thanks. I misunderstood the purpose of the API call.

**Comment by abitmore** [comment]
Created issue #584 to address the owner authorities issue discussed above.
